### DISCUSSION TITLE
What technologies and programming languages are being used by the Etsy API community?

### DISCUSSION THREAD

**Discussion by showtroylove** [etsy/open-api#951] [Q&A]
We are employing multiple languages and frameworks to target the platforms we're interested in releasing our app on.  Please share what you're using or plan to use to interact with the Etsy API.  The following is a list of technologies we're using:

### **Desktop / Mobile**

- **Target OS'**: Windows, Mac OS, iOS, Android.
- **Frameworks**:  .NET MAUI framework (frontend), WebAPI REST (backend)
- **Languages**:  C#, SQL

### **Web** 

- Frameworks: React, Node.js
- Languages: JavaScript

**Comment by Gareth064** [comment]
Hey 👋

I am a Dotnet Dev through and through. So my entire stack is Dotnet.

My current app is built using the following stack.

**Frontend**
Blazor WASM (MudBlazor UI framework)

**Backend**
ASPNET Core API web server
MS SQL

**Authentication**
Azure B2C with custom logic to handle multi tenancy

I am also building, and using, a Dotnet wrapper library for the API called **EtsyApiSharp**. It's still a private repo on GitHub but hope to make it public soon when I feel it's at a decent standard. (Well, the standard of my ability 🤣)

The plan for the API library is to be the defacto library that Any Dotnet developer would use when building something against Etsy.

**Reply by DevTKSS** [reply]
@showtroylove  @Gareth064 Nuget seems to not know your wrapper so far, github does also not find it. I am looking for anything usable to get my Authentication Code Request somehow working in c# net9... Would prefer using Refit if possible, but while I transfered the PKCE part from js to c# I am hopelessly stuck in setting up the authenticationService/Provider. #1443 maybe you would be able to help me?
using kiota with etsy's openAPI spec generated does not generate the auth endpoints at all and I found not one c# sample code for that "simple" task...

**Comment by showtroylove** [comment]
That’s awesome. Thanks for sharing.  As we’re building out our
infrastructure we’re still considering framework alternatives and Blazor
also is being considered.

Did you employ the OpenAPI .NET client code generator as a base to get you
started with the Etsy API?

**Reply by Gareth064** [reply]
Yeah I got a new lease for live when Blazor was was announced. I really dislike having to learn C# and JavaScript at the same time 🤣

The reasons I decided to write my own Library from scratch are because I didn't like the generated code that came from those tools. There were too many bugs in it which I would have to constantly "fix" every time I regenerated the code etc. Plus, I wanted the challenge of writing it from scratch. I've learned so much from doing it. Especially around how to use HTTPClient.

Btw, when you said "OpenAPI"... I read it as "OpenAI" and was confused as you would specifically ask about using AI 🤣.

Interesting though... I wonder if you could paste the OpenAPI Json co tent into ChatGPT and ask it to write a C# library class from it 🤔

**Comment by showtroylove** [comment]
Yeah, the ChatGPT is an interest new feature we've already begun
introducing into our app.  Right now we're using it to proffer suggestions
for product titles and descriptions.  It's really impressive at doing so.
  From what I've seen from ChatGPT so far - it probably could write better
code than the OpenAPI C# code generator.

The OpenAPI 3 client generator served its purpose for us.  It created the
Etsy API boilerplate code - the mundane parts anyway.  It saved us time by
writing each POCO class with all the required properties and methods.
That's all code generators have ever really been good for.  Then we
refactor each generated class as it comes into focus.  Currently we're
working on Listings.  In other words, displaying the products - going from
concept, design, mockup to publishing to one or more sales channels.

We started this venture only last month March 2023.  By my estimation, with
our current (small) team size and ambitious feature set - we have way to go
before we release a MVP.  Are you selling your app commercially right now?
If so, where are you hosting it.  It's not like Etsy has an App store like
Microsoft, Android or Apple?

**Reply by Gareth064** [reply]
With my app I am focusing on Order/Receipt Management first. It will be commercial but I'm not there yet. For now, my wife and her staff uses it.

I'm a one man band too. Wish I had a team of people 😞

Its Dev version is currently hosted on Azure services, but I am starting to realise how expensive that is getting. So once the app is live and running on Production tier services, I need to get paying customers fast 🤣

Once I get to Listing Management, I will be doing the same with GPT4, title, description and tags generation.

My app will be used to sync multiple e-commerce platforms into one place. I need to work on syncing Shopify and eBay stores next since my wife sells there as well. But 99% of her sales come from Etsy, so it's the priority.

**Comment by showtroylove** [comment]
Sounds similar to our goals.  Right now it's used internally by our team as
a POC.  We want to understand what shop owner business process (pain points
and all) is like.  This should help us make the best product.  Our vision
also unifies multiple sales channels and POD providers under one roof, if
you will.   Curious to know if you know of any other existing, generally
available products like this on the market already?

**Reply by Gareth064** [reply]
The biggest pain point with Etsy, that my app will solve for my wife, is that it is not Multi-user. Meaning her staff can see and do anything on Etsy. She especially doesn't like her staff being able to see stats and financial data. So my apps first concern was making it a multi-user experience with User Roles locking down areas of the app. The only part of my app that her staff need to see is the list of open orders so they can work on them.

Aspirations for my app are as follows but are not limited to:
- All features work with Multiple Sales Channels (create a listing/product once in my app and it pushes it to all sales channels)
- Inventory management (track individual components used to make products, as well as tracking finished products too)
- Shipping (integrate with shipping/carriers.)
- Expense & Tax tracking (preferably with bank sync to import expenses automatically)
- Integrate AI assistance as much as possible.

Examples of others that do similar things
craftybase.com
cin7.com
welcome.getvela.com

All these are great apps, but none of them do "everything" or some of the things they do are badly implemented (my subjective view), which is why I want to have a crack at building an app to attempt to do as much as I need for my wife's business to make it work as efficient as it can, but then work with any other users to implement more features that are requested.
 
Little market research tip for you. Look for things like "Roadmaps" or "Wish lists" on other apps sites. Plus join Etsy related and other apps communities on Facebook. You get a lot of insight into what users want by what they post in these places.

The biggest thing I will struggle with all of it is that I am a one man team, who is an amateur at developing and who is very weak at UI/UX stuff. So although I might get some decent features built, I think I will struggle with the implementation of them on the frontend.

**Comment by eschramm** [comment]
Recent lurker, first-time poster 👋 

I'm probably an oddity for sure. Once my wife's Etsy shop started making tax-worthy sales, and I started investigating all that entails, I quickly started working on a solution for us. So I have a hobby macOS app that we use for our shop to help us establish pricing based on raw materials' cost, labor, etc., and track sales and materials sold for taxes. 

I'm using the standard Apple native stack: Swift, Xcode, SPM, etc. And I'm using the [OAuth2](https://github.com/p2/OAuth2) library as well. The app uses the Etsy API to pull in recently shipped receipts which beats using the CSV (which kept changing all the time) and manual entry.
